SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestNetReceive/SpatialTestNetReceive.cpp

```diff
@@ -0,0 +1,185 @@
+ *before repnotifies are called on either.
+ *
+ * - Test:
+ *	- The Server changes the values of the replicated properties multiple times.
+ *  - The clients check whether the RepNotifies have or have not been called correctly.
+ * - Cleanup:
+ *	- None.
+ */
+
+static constexpr float StepTimeLimit = 15.0f;
+ASpatialTestNetReceive::ASpatialTestNetReceive()
+	: Super()
+{
+	Author = "Arthur";
+	Description = TEXT("Test Pre/PostNetReceive events");
+}
+
+void ASpatialTestNetReceive::PrepareTest()
+{
+	Super::PrepareTest();
+
+	AddStep(TEXT("SpatialTestNetReceiveInitialise"), FWorkerDefinition::Server(1), nullptr, [this]() {
+		TestActor = GetWorld()->SpawnActor<ASpatialTestNetReceiveActor>(Server1Pos, FRotator::ZeroRotator, FActorSpawnParameters());
+		TestActor->SetReplicates(true);
+
+		if (!AssertIsValid(TestActor, TEXT("Failed to spawn TestActor")))
+		{
+			return;
+		}
+
+		RegisterAutoDestroyActor(TestActor);
+
+		TestActor->Subobject->TestInt = 5;
+		TestActor->Subobject->ServerOnlyTestInt = 6;
+
+		FinishStep();
+	});
+
+	AddStep(
+		TEXT("SpatialTestNetReceiveCheckAllOk"), FWorkerDefinition::Server(2), nullptr, nullptr,
+		[this](float Dt) {
+			if (!RequireEqual_Bool(IsValid(TestActor), true, TEXT("TestActor should be valid")))
+			{
+				return;
+			}
+			if (!RequireEqual_Int(IsValid(TestActor->Subobject), true, TEXT("TestActor's subobject should be valid")))
+			{
+				return;
+			}
+
+			RequireEqual_Int(TestActor->Subobject->TestInt, 5, TEXT("TestInt property should be updated"));
+			RequireEqual_Int(TestActor->Subobject->ServerOnlyTestInt, 6, TEXT("ServerOnlyTestInt property should be updated"));
+
+			RequireEqual_Bool(TestActor->Subobject->bOnRepTestInt1Called, true, TEXT("bOnRepTestInt1Called should be the correct value"));
+			RequireEqual_Bool(TestActor->Subobject->bPreNetReceiveCalled, true, TEXT("bPreNetReceiveCalled should be the correct value"));
+			RequireEqual_Bool(TestActor->Subobject->bPostNetReceiveCalled, true, TEXT("bPostNetReceiveCalled should be the correct value"));
+
+			RequireEqual_Bool(TestActor->Subobject->bPostNetCalledBeforePreNet, false,
+							  TEXT("bPostNetCalledBeforePreNet should be the correct value"));
+			RequireEqual_Bool(TestActor->Subobject->bRepNotifyCalledBeforePreNet, false,
+							  TEXT("bRepNotifyCalledBeforePreNet should be the correct value"));
+
+			RequireEqual_Bool(TestActor->Subobject->bPreNetCalledBeforePostNet, true,
+							  TEXT("bPreNetCalledBeforePostNet should be the correct value"));
+			RequireEqual_Bool(TestActor->Subobject->bRepNotifyCalledBeforePostNet, false,
+							  TEXT("bRepNotifyCalledBeforePostNet should be the correct value"));
+
+			RequireEqual_Bool(TestActor->Subobject->bPreNetCalledBeforeRepNotify, true,
+							  TEXT("bPreNetCalledBeforeRepNotify should be the correct value"));
+			RequireEqual_Bool(TestActor->Subobject->bPostNetCalledBeforeRepNotify, true,
+							  TEXT("bPostNetCalledBeforeRepNotify should be the correct value"));
```

Comment:
This a lot of fiddly booleans, I suggest using an enum and restructuring this code.
```
enum class RepStep
{
	None,
	PreRep,
	PreNetReceive,
	PostNetReceive,
	OnRep,
}

TMap<RepStep, RepStep> RepStepToPrevRepStep;
RepStepToPrevRepStep.Add(RepStep::PreNetReceive, RepStep::None);
RepStepToPrevRepStep.Add(RepStep::PostNetReceive, RepStep::None);
RepStepToPrevRepStep.Add(RepStep::OnRep, RepStep::None);

TMap<RepStep, int32> RepStepCallCount;
RepStepCallCount.Add(RepStep::PreNetReceive, 0);
RepStepCallCount.Add(RepStep::PostNetReceive, 0);
RepStepCallCount.Add(RepStep::OnRep, 0);
RepStep CurrentStep = RepStep::PreRep;

RequireEqual_Bool(RepStepToPrevRepStep[RepStep::PostNetReceive] == RepStep::PreNetReceive, true,
							  TEXT("bPostNetCalledBeforePreNet should be the correct value"));

void USpatialTestNetReceiveSubobject::PreNetReceive()
{
	RepStepToPrevRepStep[RepStep::PreNetReceive] = CurrentStep;
	RepStepCallCount[RepStep::PreNetReceive] = ++RepStepCallCount[RepStep::PreNetReceive];
}
... etc etc

Reply:
Done something very similar to this. Thanks for the detail, sped up the implementation process